src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala

```diff
@@ -244,7 +244,10 @@ case class MergeIntoCommand(
 
     // Calculate frequency of matches per source row
     val matchedRowCounts = collectTouchedFiles.groupBy(ROW_ID_COL).agg(sum("one").as("count"))
-    if (matchedRowCounts.filter("count > 1").count() != 0) {
+    // Even there is no update clause, we still need this inner join job to fill touchedFilesAccum,
+    // but the exception is thrown only with update clause.
+    if (matchedRowCounts.filter("count > 1").count() != 0 &&
+        (updateClause.isDefined || !isMatchedOnly)) {
```

Comment:
why isn't this just an `updateClause.isDefined`?

Reply:
@LantaoJin any response to this? we would like to merge this. I think this is a valid improvement.

Reply:
Ok, I will review this part and reply soon. I need to reload the memory :)

Reply:
Any update on this @LantaoJin ?

Reply:
Sorry for replying late. I will update the code tomorrow. I am upgrading our internal Delta to 0.7 and I find some bugs. I will file issues and PRs soon.

Reply:
We are planning to cut a release soon and would be good to have this. In fact, I can take care of making small changes if you just give us the reason why `!isMatchedOnly` is needed?

Reply:
Ah, I see, `!isMatchedOnly` is to make the unit test "insert only merge - turn off feature flag" pass. I am not sure why the UT expects an exception. But to minimize the impact, I added `!isMatchedOnly`

Reply:
If `! isMatchedOnly ` removed, the UT will change to
```scala
  test("insert only merge - turn off feature flag") {
    withSQLConf(DeltaSQLConf.MERGE_INSERT_ONLY_ENABLED.key -> "false") {
      withKeyValueData(
        source = (0, 0) :: (1, 10) :: (1, 100) :: (3, 30) :: (3, 300) :: Nil,
        target = (1, 1) :: (2, 2) :: Nil
      ) { case (sourceName, targetName) =>
        // This is supposed to fail as the duplicated keys in source were not supported.
        executeMerge(s"$targetName t", s"$sourceName s", "s.key = t.key",
          insert(values = "(key, value) VALUES (s.key, s.value)"))
        checkAnswer(
          readDeltaTable(tempPath),
          Seq(
            Row(0, 0),      // Existing value
            Row(3, 300),    // Existing value
            Row(3, 30),     // Existing value
            Row(1, 1),      // Insert and replace existing value
            Row(1, 1),      // Insert and replace existing value
            Row(2, 2),      // Insert
          )
        )
      }
    }
  }
```
Is this what we want? @tdas

Reply:
Aah interesting. Actually this is a bit more complicated than I thought. Here are my thoughts. 

1. This test failure shows flipping off insert-only-merge optimization causes incorrect results because if use the common code path (that uses 2 joins) and not throw error on multiple match, then the second join (in `writeAllChanges` will cause spurious joined rows from the multiple matches and generate extra rows (e.g., the above case produces two (1, 1) from the two matches with source data (1, 10) and (1, 100)). So while restricting the optimization with !isMatchedOnly is correct, it suboptimal because we are not allowing unambiguous cases like `unconditional delete + unconditional insert`. 

2. Furthermore, it made me think and realize that condition `(updateClause.isDefined || !isMatchedOnly)` is wrong in a different way. `conditional delete` (e.g. `WHEN MATCHED AND <condition> THEN DELETE`) should throw also error as well, as multiple matches that may or may not satisfy the delete clause condition and therefore is ambiguous.

Reply:
I have experimented with this a little bit by writing more tests. I think here is the right thing to do. 

1. We should make sure that the main, non-optimized code path using 2 joins (`findTouchedFiles` and `writeAllChanges`) always works correctly for all cases, including insert-only merge when the insert-only-optimization is disabled. In addition, we want to capture more unambiguous cases like unconditional delete + unconditional insert, where as reject cases which are either ambiguous (conditional delete) or cannot be computed correctly (insert-only when the optimization is disabled).

2. This basically means we have to be more nuanced in our condition. Here is what I came up with. 

```
    // Get multiple matches and simultaneously collect (using touchedFilesAccum) the file names
    val multipleMatchCount = matchedRowCounts.filter("count > 1").count()

    // Throw error if multiple matches are ambiguous
    val isAmbiguous = {
      // Multiple matches are not ambiguous only when there is only one unconditional delete as
      // all the matched row pairs in the 2nd join (see `writeAllChanges`) will get deleted.
      val isUnconditionalDelete = matchedClauses.headOption match {
        case Some(DeltaMergeIntoDeleteClause(None)) => true
        case _ => false
      }
      !(matchedClauses.size == 1 && isUnconditionalDelete)
    }

    if (multipleMatchCount > 0 && isAmbiguous) { 
     ... // throw eror
    }
```
What do you think?

Furthermore, in the interest of time, do you mind if I take over this PR? Of course, either way, you get authorship.

Reply:
Correct. Current the MergeInto is much more complicated than what I am using. I leverage MergeInto to implement the cross tables delete and cross tables update, such as:
```sql
UPDATE t
FROM A t, B s
SET t.col = s.col
Where t.id=s.id AND s.col != 9
```
and 
```sql
DELETE t
FROM A t, B s
Where t.id=s.id AND s.col != 9
```

Reply:
>conditional delete (e.g. WHEN MATCHED AND <condition> THEN DELETE) should throw also error as well

So how about
```scala
    if (matchedRowCounts.filter("count > 1").count() != 0 &&
        (updateClause.isDefined || !isMatchedOnly ||
          (deleteClause.isDefined && updateClause.exists(_.condition.isDefined)))) {
      throw DeltaErrors.multipleSourceRowMatchingTargetRowInMergeException(spark)
    }
```

Reply:
Sorry, I missed your second comment before previous replying. I will update the code.